src/rgw/rgw_kafka.cc

```diff
@@ -194,6 +196,8 @@ void log_callback(const rd_kafka_t* rk, int level, const char *fac, const char *
 void poll_err_callback(rd_kafka_t *rk, int err, const char *reason, void *opaque) {
   const auto conn = reinterpret_cast<connection_t*>(rd_kafka_opaque(rk));
   ldout(conn->cct, 10) << "Kafka run: poll error(" << err << "): " << reason << dendl;
+  conn->status = err;
```

Comment:
so from what i understand is this `poll_err_callback` will return an error for when the kafka is down and now the connection->status is set to `ERROR` returned from this function.
Now the next call to same topic on non-persistent notification will call `is_alive` and that will retrieve the connection from the map and return the status and since the status is not OK, it will not even try to send the notification? And then eventually the notification will be removed from the map after the `IDLE_TIMEOUT` and now next time create_endpoint will try re-creating the connection and store it in the map.
Am i correct in understanding this workflow ?

Reply:
> so from what i understand is this `poll_err_callback` will return an error for when the kafka is down and now the connection->status is set to `ERROR` returned from this function. Now the next call to same topic on non-persistent notification will call `is_alive` and that will retrieve the connection from the map and return the status and since the status is not OK, it will not even try to send the notification? 

correct. not only it will not send the notification, it will fail the operation (PUT, DELETE etc.) that triggered the notification.

> And then eventually the notification will be removed from the map after the `IDLE_TIMEOUT` 

the connection will be removed from the map after the `IDLE_TIMEOUT` if there are no ***attempts*** to send to it, regardless of the status of the connection. this is to do cleanup of decomissioned brokers, and cleanup if all topics that point to a brokers are deleted. and is unrelated to the state of the conection

> and now next time create_endpoint will try re-creating the connection and store it in the map. Am i correct in understanding this workflow ?

if the broker is down (in error state) it is not deleted, we retry to conect in the main thread. see:
https://github.com/ceph/ceph/blob/8214dfa5ca514991bac86a2e010bc8d93c65fd4d/src/rgw/rgw_kafka.cc#L480

Reply:
> if the broker is down (in error state) it is not deleted, we retry to conect in the main thread. see:

the main thread just resets all the config param of kafka and also sets the `STATUS = OK`.
this essentially will lead to non deterministic behavior, Some S3(PUT,DELETE,etc) request will receive 200 but notification event will have FAILED due to broker down and some S3 request that sneak in between the window where the `poll_err_callback` sets the status to ERROR and `main::run` sets is back to `OK`, will receive `503/EINVAL`.
what i would suggest is, let each S3 request try to send the notification event to broker. The `message_callback` or  `poll_err_callback` will receive the error from broker immediately and the just invoke the s3 callback for error status. currently only `message_callback` invokes the callback, we can invoke the s3:callback from `poll_err_callback` as well? Essentially the issue we are trying to solve here is not wait till IDLE_TIMEOUT for the status to reflect ? correct?

Reply:
> > > if the broker is down (in error state) it is not deleted, we retry to conect in the main thread. see:
> > 
> > 
> > the main thread just resets all the config param of kafka and also sets the `STATUS = OK`. this essentially will lead to non deterministic behavior, Some S3(PUT,DELETE,etc) request will receive 200 but notification event will have FAILED due to broker down and some S3 request that sneak in between the window where the `poll_err_callback` sets the status to ERROR and `main::run` sets is back to `OK`, will receive `503/EINVAL`.
> 
> you are right
> 
> > what i would suggest is, let each S3 request try to send the notification event to broker. The `message_callback` or `poll_err_callback` will receive the error from broker immediately and the just invoke the s3 callback for error status.
> 
> `poll_err_callback` is invoked from `rd_kafka_poll` and cannot be invoked immediatly or per messsage.
> 
> > currently only `message_callback` invokes the callback, we can invoke the s3:callback from `poll_err_callback` as well?
> 
> no
> 
> > Essentially the issue we are trying to solve here is not wait till IDLE_TIMEOUT for the status to reflect ? correct?
> 
> `IDLE_TIMEOUT` is a different and unrelated issue. is was _**wrongly**_ used before as a way to clean up error states.
> 
> one thing we can do to be more deterministic is to send a dummy message and then poll for an answer every time we create a new connection. and set the status accordingly. it will take longer to create a connection, but since we reuse the connection when sending notifications it should not really impact overall perf.

sending a dummy message will be same as current, dummy message would succeed at start but assuming the broker goes down in the middle, then all the events sent in middle will still not not delivered.
reading through the kafka doc, can we leverage this `rd_kafka_conf_set_background_event_cb`, this gets called immediately once the broker is down and then we could go and update the status of connection for that topic ?

But one thing is sure, to have consistent behavior we do not have to reject/error the s3 request if the broker is down for `NON Persistent` notification, and continue the best effort logic.

Reply:
> sending a dummy message will be same as current, dummy message would succeed at start but assuming the broker goes down in the middle, then all the events sent in middle will still not not delivered.

IMO, there is a big difference between sending the wrong response while the broker goes down (underiable, but acceptble) and sending wrong responses while the broker is down (unacceptable)

> reading through the kafka doc, can we leverage this rd_kafka_conf_set_background_event_cb, this gets called immediately once the broker is down and then we could go and update the status of connection for that topic ?

I need to test it first, but on rhel8 we use `librdkafka-devel 0.11.4` which does not support that

> But one thing is sure, to have consistent behavior we do not have to reject/error the s3 request if the broker is down for NON Persistent notification, and continue the best effort logic.

if you are good with best effort behavior, you can just set the `ack-level` to `none`. in this case the notification will fail only if there is something wrong with the configuration of the broker.